op-node/rollup/finality/plasma.go

```diff
@@ -0,0 +1,50 @@
+package finality
+
+import (
+	"context"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	plasma "github.com/ethereum-optimism/optimism/op-plasma"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+type PlasmaBackend interface {
+	// Finalize notifies the L1 finalized head so plasma finality is always behind L1.
+	Finalize(ref eth.L1BlockRef)
+	// OnFinalizedHeadSignal sets the engine finalization signal callback.
+	OnFinalizedHeadSignal(f plasma.HeadSignalFn)
+}
+
+// PlasmaFinalizer is a special type of Finalizer, wrapping a regular Finalizer,
+// but overriding the finality signal handling:
+// it proxies L1 finality signals to the plasma backend,
+// and relies on the backend to then signal when finality is really applicable.
+type PlasmaFinalizer struct {
+	*Finalizer
+	backend PlasmaBackend
+}
+
+func NewPlasmaFinalizer(log log.Logger, cfg *rollup.Config,
+	l1Fetcher FinalizerL1Interface, ec FinalizerEngine,
+	backend PlasmaBackend) *PlasmaFinalizer {
+
+	inner := NewFinalizer(log, cfg, l1Fetcher, ec)
+
+	// In plasma mode, the finalization signal is proxied through the plasma manager.
+	// Finality signal will come from the DA contract or L1 finality whichever is last.
+	// The plasma module will then call the inner.Finalize function when applicable.
+	backend.OnFinalizedHeadSignal(func(ref eth.L1BlockRef) {
```

Comment:
For later, I think the awkwardness here is due to the fact that how we track safety for plasma mode is not accurate. Long term we should add a field to the safety/finality data which is `Optimistic L1 Derivation Block` for Plasma mode which is the L1 origin when the block was derived (roughly the L1 inclusion block of the commitment). Then later we should mark it as fully safe when the commitment can no longer be challenged. That would set the `L1 Derivation block`. Then we can apply finality as normal (modulo the large lookback period).


Right now we shove the `Optimistic L1 Derivation block` into the safe block for the finality data & have to delay the L1 finalization signal until the block in which a commitment becomes no longer able to be challenged becomes finalized.

Reply:
Opened an issue to track: https://github.com/ethereum-optimism/optimism/issues/10723